### PR TITLE
fix: isolate target encoding fold priors

### LINKED ISSUE #36: TargetEncoder's out-of-fold encoding leaks through a shared global prior
`TargetEncoder.fit_transform` is supposed to be leakage-free: "each row is encoded from
the folds that do not contain it -- so a model trained on the output cannot see its own
label through the encoding." The per-fold *counts* really are restricted to the other
folds. The smoothing *prior* mixed into every fold's estimate is not:

```python
# _fit, line 634 -- computed once, over the WHOLE training target
self.prior_ = float(np.nanmean(target)) if np.isfinite(target).any() else 0.0
...
# _fit_transform, per fold, line 672-673
mapping = _smoothed_means(series[rest], target[rest], self.prior_, self.smoothing_)
```

`_smoothed_means` computes `(sum + prior * smoothing) / (count + smoothing)`. `sum` and
`count` are correctly restricted to `rest` (the other folds). `prior` is `self.prior_`,
computed from every row -- including every row in the current holdout fold. For a
category with few or zero occurrences in `rest` (a near-singleton category, exactly the
case this class exists to protect), the encoded value collapses toward `self.prior_`,
which was partly computed from that row's own target.

### Reproduced

```python
df = pd.DataFrame({"cat": ["a", "b", "c", "d", "e"], "y": [0, 0, 0, 0, 0]})
# 5 singleton categories -> every row's OOF value falls back entirely on self.prior_
TargetEncoder(["cat"], n_folds=5, random_state=0).fit_transform(df[["cat"]], df["y"], ctx)["cat"]
# [0.0, 0.0, 0.0, 0.0, 0.0]

df2 = df.copy(); df2["y"] = [1, 0, 0, 0, 0]  # flip ONLY row 0's own label
TargetEncoder(["cat"], n_folds=5, random_state=0).fit_transform(df2[["cat"]], df2["y"], ctx2)["cat"]
# [0.2, 0.2, 0.2, 0.2, 0.2]
```

Row 0's own out-of-fold encoded value changes purely because row 0's own target
changed -- the exact failure mode the docstring says this class prevents.

Why the existing leakage test didn't catch it: `test_target_encoding_is_cross_fitted`
(`tests/test_leakage.py:194`) only asserts an aggregate correlation
(`corr_oof < 0.2`) on a frame where most categories have decent support. The leak is
strongest exactly where support is thinnest -- rare/near-singleton categories -- which a
single aggregate correlation over a whole synthetic frame can easily average away.

### What to do

Compute the prior per fold, from `target[rest]` only, rather than reusing the
already-computed `self.prior_`:

```python
fold_prior = float(np.nanmean(target[rest])) if np.isfinite(target[rest]).any() else self.prior_
mapping = _smoothed_means(series[rest], target[rest], fold_prior, self.smoothing_)
...
out[holdout] = encoded.fillna(fold_prior).to_numpy(dtype=np.float64)
```

`self.prior_` (the full-train prior) stays correct and necessary for `_transform` on
genuinely new data at prediction time -- it's only the `_fit_transform` per-fold path
that needs its own, fold-restricted prior.

### Tests

- Extend or add alongside `test_target_encoding_is_cross_fitted`: a frame where every
  category is a near-singleton (as in the repro above), asserting that flipping one
  row's own target does not change that row's own OOF-encoded value.
- `test_target_encoder_fit_then_transform_differs_from_fit_transform` already exists;
  worth checking it still passes and still demonstrates a real difference after the fix.

**Before you start:** please comment to claim this issue, and check whether someone
already has.

### PR DESCRIPTION
**Pull request by XonkelX**
## Summary

`TargetEncoder.fit_transform()` reused the full-training prior while building out-of-fold encodings, allowing each holdout target to influence its own smoothed value or unseen-category fallback. This change computes the prior from each fold's training partition and uses it for both smoothing and holdout fallback. The fitted full-training prior remains unchanged for normal `transform()` calls.

A singleton-category regression test flips every row's target in turn and verifies that the row's own OOF encoding does not change. The fold-local calculation also explicitly falls back to the full-training prior when a training partition contains no finite target values.

Closes #36

## Validation

- `pytest` — 403 passed
- `ruff check src/ tests/ benchmarks/ examples/` — passed

### CODE REVIEW

**Comment by bijay-odyssey** [comment]
Merged in 8d61c21. This was the most severe finding out of the whole audit — the library's actual central promise ("no leakage, structurally") had a real gap in it, present in every release since 0.1.0 — and you got the fix exactly right on the first pass, including both of the sharp edges I flagged (the `.fillna()` side, and writing the all-NaN-fold fallback down explicitly instead of leaving it implicit).

I ran my own independent repro before merging, past your test: flipping only row 0's own label now leaves row 0's own OOF value unchanged, while the *other* rows correctly still shift — which is exactly the distinction that matters (a row's own encoding must not depend on its own label, but it's supposed to depend on other rows' labels when they land in its training folds). Good catch keeping that boundary intact rather than over-correcting.

Cutting a release for this promptly given what it fixes — will link it here once it's out.
